segmentstore/server/src/main/java/io/pravega/service/server/logs/DurableLog.java

```diff
@@ -234,8 +235,11 @@ public int getId() {
         TimeoutTimer timer = new TimeoutTimer(timeout);
         log.info("{}: Truncate (OperationSequenceNumber = {}, DataFrameAddress = {}).", this.traceObjectId, upToSequenceNumber, truncationFrameAddress);
 
-        return this.durableDataLog
-                .truncate(truncationFrameAddress, timer.getRemaining())
+        // Before we do any real truncation, we need to mini-snapshot the metadata with only those fields that are updated
+        // asynchronously for us (i.e., not via normal Log Operations) such as the Storage State. That ensures that this
+        // info will be readily available upon recovery without delay.
+        return add(new StorageMetadataCheckpointOperation(), timer.getRemaining())
```

Comment:
1. MetadataCheckpoin (MC) vs StorageMetdataCheckpoint (SMC)
    - MC checkpoints **everything**. That includes metadata updated via normal operations (length, sealed, merge status, etc.). It also records the mapping of Segment Id to Segment Names. As such, for 10K active segments, it can get large (1-2MB avg, but for very long segment names it could get up to 10MB). As such, I don't want that queued up every time we do a truncate. MCs are added periodically (every 100-200MB) based on configuration - they essentially define where we can truncate Tier1.
    - SMC is a slimmer version of that, where it only maps segment id, storage length and sealed status (in storage). Even for 10K segments, it will not exceed 170KB (less with compression). That was the main reasoning behind having another type of checkpoint. 
    - A Container recovery can only start from a full MC (which is essentially a summary of all operations up until it); subsequent MCs are ignored (since they're redundant), but every time a SMC is encountered, its contents is applied.
2. I think I managed to answer this question above. Yes, we could (I could just put MCs 
everywhere), but those tend to be big and kinda redundant, so I want something smaller.
3. What I meant by "asynchronously" was "not by the `OperationMetadataUpdater` (OMC). The OMC is the only thing updating the metadata (since it's part of the `OperationProcessor`) and it essentially runs in a single thread (ok, across many threads in a thread pool, but it runs in sequence). Async is the `Writer` updating `StorageLength` and `StorageSealed` as it applies those changes to Tier2. 
4. Check out `DurableLog.truncate`. It queues up a SMC, then after it completes it issues the Tier1 truncation.